### PR TITLE
Add optional reference point to problem

### PR DESCRIPTION
**Pull request by hvy**
Introduces a reference point field `reference_point` to Kurobako's problems. Prior to this change, the reference points could be specified via `attrs` which was somewhat ad-hoc. This PR also adds an actual reference point to ZDT for demonstrational purpose.

This field is 

- read from when plotting hypervolumes.
- is optional. If `None`, plotting will fall back to a hard-coded point.

### CODE REVIEW

File: kurobako_problems/src/zdt.rs [hunk 1]
```diff
@@ -54,7 +54,7 @@ impl ProblemFactory for ZdtProblemFactory {
                 "paper",
                 "Zitzler, Eckart, Kalyanmoy Deb, and Lothar Thiele. \"Comparison of multiobjective \
                  evolutionary algorithms: Empirical results.\" Evolutionary computation 8.2 (2000): 173-195."
-            ).value(domain::var("f1")).value(domain::var("f2"));
+            ).value(domain::var("f1")).value(domain::var("f2")).reference_point(Some(Params::new(vec![11.0, 11.0])));
```
Comment (sile) [inline kurobako_problems/src/zdt.rs:57]: Just out of curiosity. Could you tell me where the reference point `[11.0, 11.0]` came from?

File: kurobako_core/src/problem.rs [hunk 2]
```diff
@@ -64,6 +66,12 @@ impl ProblemSpecBuilder {
...
     }
 
+    /// Sets the value reference point of this problem.
+    pub fn reference_point(mut self, reference_point: Option<Params>) -> Self {
+        self.reference_point = reference_point;
+        self
+    }
+
     /// Builds a `ProblemSpec` with the given settings.
     pub fn finish(self) -> Result<ProblemSpec> {
```
Comment (sile) [inline kurobako_core/src/problem.rs:76]: I think that we should check whether the dimension of `self.values` matches the dimension of `self.reference_point` here (if the latter isn't `None`).

Comment (hvy) [reply kurobako_core/src/problem.rs:76]: Makes sense. Defined a private function `validate` returning a `Result<()>` in https://github.com/sile/kurobako/pull/33/commits/84ee3f22c28f4ee5c9ea9814b2f58e89d99bad1d. I'm not so sure whether I'm doing something ugly or anti-pattern. Glad for any sorts of feedback. Thanks.

File: kurobako_problems/src/zdt.rs (hunk 1 above)
Comment (hvy) [reply kurobako_problems/src/zdt.rs:57]: It's based on [BoTorch's test set](https://github.com/pytorch/botorch/blob/master/botorch/test_functions/multi_objective.py). Might be worth noting that BoTorch does not provide the full set of ZDT functions as Kurobako does. I observed that we could get strange hypervolumes with e.g. ZDT4.

**Comment by hvy** [comment]
PTAL.

File: kurobako_problems/src/zdt.rs (hunk 1 above)
Comment (sile) [reply kurobako_problems/src/zdt.rs:57]: Got it. Thanks!

File: kurobako_core/src/problem.rs [hunk 3]
```diff
@@ -76,8 +86,23 @@ impl ProblemSpecBuilder {
...
+    fn validate(&self) -> Result<()> {
+        if let Some(reference_point) = &self.reference_point {
+            if reference_point.len() != self.values.len() {
+                track_panic!(
+                    ErrorKind::InvalidInput,
+                    "Unexpected dimensions of values {} and reference point {}",
+                    self.values.len(),
+                    reference_point.len()
+                )
+            }
```
Comment (sile) [inline kurobako_core/src/problem.rs]: We can use `track_assert_eq!` as follows:
```suggestion
            track_assert_eq!(
                reference_point.len(),
                self.values.len(),
                ErrorKind::InvalidInput,
                "Unexpected dimensions of values {} and reference point {}",
                self.values.len(),
                reference_point.len()
            )
```

File: kurobako_core/src/problem.rs [hunk 4]
```diff
@@ -64,8 +66,16 @@ impl ProblemSpecBuilder {
...
 
+    /// Sets the value reference point of this problem.
+    pub fn reference_point(mut self, reference_point: Option<Params>) -> Self {
+        self.reference_point = reference_point;
+        self
+    }
+
     /// Builds a `ProblemSpec` with the given settings.
     pub fn finish(self) -> Result<ProblemSpec> {
+        self.validate()?;
```
Comment (sile) [inline kurobako_core/src/problem.rs]: Please wrap this expression by the `track!` macro (it adds trace information when an error occurs).
```suggestion
        track!(self.validate())?;
```

**Review by sile — COMMENTED** [review]
Thank you for your update!
I left two nitpicking comments 🙏

**Comment by hvy** [comment]
Thanks for your suggestions, applied and PTAL.

**Comment by hvy** [comment]
Codecov upload failure by CI.

```  
with:
    github-token: ***
    parallel: true
    path-to-lcov: /tmp/grcov-report-vrio2uc3gci
    coveralls-endpoint: https://coveralls.io
Using lcov file: /tmp/grcov-report-vrio2uc3gci
Error:  "2021-01-11T01:04:41.932Z"  'error from getOptions'
Error: Error: Command failed: git cat-file -p f7c81312770f389e4b17f8b8f63f879a9e55324b
fatal: Not a valid object name f7c81312770f389e4b17f8b8f63f879a9e55324b
```

Looks similar to the "race condition" issue with `checkout@v1`  but I'm not entirely sure since `master` haven't been updated. How about trying `checkout@v2`?

**Comment by sile** [comment]
> How about trying checkout@v2?

Sounds good. Thank you for your suggestion.

**Review by sile — APPROVED** [review]
LGTM. Thank you for your contribution!
(I'll try checkout@v2 later)

**Pull request merged**